File: ServiceBookingSystem/ServiceBookingSystem/src/main/java/com/watuka/ServiceBookingSystem/Services/Client/ClientService.java
package com.watuka.ServiceBookingSystem.Services.Client;

import com.watuka.ServiceBookingSystem.Dto.AdDTO;
import com.watuka.ServiceBookingSystem.Dto.AdDetailsForClientDTO;
import com.watuka.ServiceBookingSystem.Dto.ReservationDTO;
import com.watuka.ServiceBookingSystem.Dto.ReviewDTO;

import java.util.List;

public interface ClientService {

    List<AdDTO> getAllAds();

    List<AdDTO> searchAdByName(String name);

    boolean bookService(ReservationDTO reservationDTO);

    AdDetailsForClientDTO getAdDetailsByAdId (Long adId);

    List<ReservationDTO> getAllBookingsByUserId(Long userId);

    Boolean giveReview(ReviewDTO reviewDTO);
}

File: serviceBookingFront/src/app/client/services/client.service.ts
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { UserStorageService } from 'src/app/basic/services/storage/user-storage.service';

const BASIC_URL = "http://localhost:8080/";

@Injectable({
  providedIn: 'root'
})
export class ClientService {

  constructor(private http: HttpClient) { }

  getAllAds(): Observable<any> {
    return this.http.get(BASIC_URL + `api/client/ads`, {
      headers: this.createAuthorizationHeader()
    })
  }

  searchByName(name: any): Observable<any> {
    return this.http.get(BASIC_URL + `api/client/search/${name}`, {
      headers: this.createAuthorizationHeader()
    })
  }

  getAdDetailsById(adId: any): Observable<any> {
    return this.http.get(BASIC_URL + `api/client/ad/details/${adId}`, {
      headers: this.createAuthorizationHeader()
    })
  }

  bookService(bookDTO: any): Observable<any> {
    return this.http.post(BASIC_URL + `api/client/book-service`, bookDTO, {
      headers: this.createAuthorizationHeader()
    })
  }

  getMyBookings(): Observable<any> {
    const userId = UserStorageService.getUserId();
    return this.http.get(BASIC_URL + `api/client/my-bookings/${userId}`,{
      headers: this.createAuthorizationHeader()
    })
  }

  giveReview(reviewDTO: any): Observable<any> {
    return this.http.post(BASIC_URL + `api/client/review`, reviewDTO, {
      headers: this.createAuthorizationHeader()
    })
  }


  createAuthorizationHeader(): HttpHeaders{
    const token = UserStorageService.getToken();
    let authHeaders = new HttpHeaders();

    if(token){
      return authHeaders.set('Authorization', 'Bearer ' + token)
    }else{
      console.error('Token Missing !');
      return authHeaders;
    }
  }
}
